pkg/ccl/backupccl/targets.go

```diff
@@ -321,15 +321,96 @@ func fullClusterTargetsBackup(
 	return fullClusterDescs, fullClusterDBIDs, nil
 }
 
+// checkMissingIntroducedSpans asserts that each backup's IntroducedSpans
+// contain all tables that require an introduction (i.e. a backup from ts=0), if
+// they are also restore targets. This specifically entails checking the
+// following invariant on each table we seek to restore: if the nth backup
+// contains a table in its backupManifest.Descriptors field but the n-1th does
+// not, then that table must be covered by the nth backup's
+// manifest.IntroducedSpans field. This check only applies to backups with a
+// start time that is less than the restore AOST.
+func checkMissingIntroducedSpans(
+	restoringDescs []catalog.Descriptor,
+	mainBackupManifests []BackupManifest,
+	endTime hlc.Timestamp,
+	codec keys.SQLCodec,
+) error {
+	// Gather the tables we'd like to restore.
+	requiredTables := make(map[descpb.ID]struct{})
+	for _, restoringDesc := range restoringDescs {
+		if _, ok := restoringDesc.(catalog.TableDescriptor); ok {
+			requiredTables[restoringDesc.GetID()] = struct{}{}
+		}
+	}
+
+	for i, b := range mainBackupManifests {
+		if !endTime.IsEmpty() && endTime.Less(b.StartTime) {
+			// No need to check a backup that began after the restore AOST.
+			return nil
+		}
+
+		if i == 0 {
+			// The full backup does not contain reintroduced spans
+			continue
+		}
+
+		// Gather the tables included in the previous backup
+		prevTables := make(map[descpb.ID]struct{})
+		for _, desc := range mainBackupManifests[i-1].Descriptors {
+			if table, _, _, _ := descpb.FromDescriptor(&desc); table != nil {
+				prevTables[table.GetID()] = struct{}{}
+			}
+		}
+
+		// Gather the introduced tables
+		tablesIntroduced := make(map[descpb.ID]struct{})
+		for _, span := range mainBackupManifests[i].IntroducedSpans {
+			_, tableID, err := codec.DecodeTablePrefix(span.Key)
+			if err != nil {
+				return err
+			}
+			tablesIntroduced[descpb.ID(tableID)] = struct{}{}
+		}
+
+		// If the table in the current backup was not in prevBackup.Descriptors,
+		// then it needs to be introduced.
+		for _, desc := range mainBackupManifests[i].Descriptors {
+			if table, _, _, _ := descpb.FromDescriptor(&desc); table != nil {
+
+				if _, required := requiredTables[table.GetID()]; !required {
+					continue
+				}
+
+				_, inPrevBackup := prevTables[table.GetID()]
+
+				_, introduced := tablesIntroduced[table.GetID()]
+
+				if !inPrevBackup && !introduced {
+					tableError := errors.Newf("table %v was not properly backed up after its import"+
+						" completed. To continue the restore, "+
+						"either 1) restore from a different full backup; 2) restore to a system time before"+
+						" the import completed; 3) or remove %v from the restore targets.",
+						table.Name, table.Name)
+					return errors.WithIssueLink(tableError, errors.IssueLink{
```

Comment:
Lets do:

errors.WithIssueLink(errors.WithHint(....), ....) and add your steps to restore this table as a hint instead?

Also I don't feel great about saying: `A backup chain with revision history can lose data on a table with an import`.

I'd qualify the statement a little more: An incremental database backup with revision history can incorrectly backup data for a table that was running an IMPORT at the time of the previous incremental in this chain of backups.

Reply:
I think using `errors.WithIssueLink(errors.WithHint(....), ....)` would look kinda weird. As my data driven test indicates below, errors.WIthIssueLink already issues a `HINT` prefix. So i think your suggestion would create two hints, which seems confusing.

Reply:
The extra `HINT` might be an artifact of what we're doing here - https://github.com/cockroachdb/cockroach/blob/fe4f495d71394647ace18c475c694a7f82ba5a61/pkg/ccl/backupccl/datadriven_test.go#L584. I can try generating a corrupt backup and then restoring it with your patch and seeing how it looks in the terminal.

Reply:
aside: why change how the error msg presents in the dd test?

Reply:
Not worrying about this for now.